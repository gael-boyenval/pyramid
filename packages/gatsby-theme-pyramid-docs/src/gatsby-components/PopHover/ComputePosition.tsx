interface Position {
  right: number;
  top: number;
  bottom: number;
  left: number;
  width: number;
  height: number;
}

export const computePosition = (
  pos: Position,
  windowHeight: number,
  windowWidth: number,
) => ({
  top: {
    global: {
      bottom: windowHeight - pos.bottom,
      left: pos.right - pos.width / 2,
      paddingBottom: pos.height,
      transform: 'translateX(-50%)',
    },
    ghost: {
      bottom: 0,
      left: '50%',
      transform: 'translateX(-50%)',
      width: pos.width + 30,
      height: pos.height + 20,
    },
  },
  right: {
    global: {
      top: pos.bottom - pos.height / 2,
      transform: 'translateY(-50%)',
      left: pos.left,
      paddingLeft: pos.width,
    },
    ghost: {
      top: '50%',
      left: 0,
      transform: 'translateY(-50%)',
      width: pos.width + 30,
      height: pos.height + 20,
    },
  },
  left: {
    global: {
      top: pos.bottom - pos.height / 2,
      transform: 'translateY(-50%)',
      right: windowWidth - pos.right,
      paddingRight: pos.width,
    },
    ghost: {
      top: '50%',
      right: 0,
      transform: 'translateY(-50%)',
      width: pos.width + 30,
      height: pos.height + 20,
    },
  },
  bottom: {
    global: {
      top: pos.top,
      left: pos.left + pos.width / 2,
      paddingTop: pos.height,
      transform: 'translateX(-50%)',
    },
    ghost: {
      left: '50%',
      top: 0,
      transform: 'translateX(-50%)',
      width: pos.width + 20,
      height: pos.height + 40,
    },
  },
});

type side = 'top' | 'left' | 'right' | 'bottom';

export const defineBestSide = (
  pos: Position,
  windowHeight: number,
  windowWidth: number,
): side => {
  const realEstate = {
    top: pos.top,
    left: pos.left,
    bottom: windowHeight - pos.bottom,
    right: windowWidth - pos.right,
  };

  const bestSide: side = ['top', 'right', 'left', 'bottom'].reduce(
    (current: side, next: side): side =>
      realEstate[current] > realEstate[next] ? current : next,
  );

  return bestSide;
};
