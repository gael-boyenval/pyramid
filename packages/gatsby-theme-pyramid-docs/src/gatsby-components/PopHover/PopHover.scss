@import '../../theme/s.theme.scss';

.c-pop-hover-wrapper {
  display: inline-flex;
  margin: 0;
}

.c-pop-hover-root {
  pointer-events: none;
  position: fixed;
  overflow: visible;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
}

@function get($key) {
  @return map-get($pop-hover, $key);
}

.c-pop-hover {
  pointer-events: none;
  position: absolute;
  overflow: visible;
  z-index: 2;
  display: inline-block;

  &__tooltip {
    position: relative;
    pointer-events: initial;

    // arrow layout
    &:after,
    &:before {
      content: '';
      z-index: 1;
      display: block;
      position: absolute;
      border-style: solid;
      height: 0px;
      width: 0px;
    }

    &:after {
      box-shadow: none;
      z-index: 2;
    }
  }

  // arrow positions variations
  &--top &__tooltip {
    padding-bottom: get('arrow-size') + get('space-between-arrow-and-trigger');

    &:after,
    &:before {
      border-color: get('bdr-color') transparent;
      border-width: get('arrow-size') get('arrow-size') 0px get('arrow-size');
      bottom: get('space-between-arrow-and-trigger');
      left: 50%;
      transform: translateX(-50%);
    }

    &:after {
      bottom: get('space-between-arrow-and-trigger') + get('bdr-width');
      border-color: get('bg') transparent;
    }
  }

  &--bottom &__tooltip {
    padding-top: get('arrow-size') + get('space-between-arrow-and-trigger');

    &:after,
    &:before {
      border-color: get('bdr-color') transparent;
      border-width: 0px get('arrow-size') get('arrow-size') get('arrow-size');
      top: get('space-between-arrow-and-trigger');
      left: 50%;
      transform: translateX(-50%);
    }

    &:after {
      top: get('space-between-arrow-and-trigger') + get('bdr-width');
      border-color: get('bg') transparent;
    }
  }

  &--right &__tooltip {
    padding-left: get('arrow-size') + get('space-between-arrow-and-trigger');

    &:after,
    &:before {
      border-color: transparent get('bdr-color');
      border-width: get('arrow-size') get('arrow-size') get('arrow-size') 0px;
      left: get('space-between-arrow-and-trigger');
      top: 50%;
      transform: translateY(-50%);
    }

    &:after {
      left: get('space-between-arrow-and-trigger') + get('bdr-width');
      border-color: transparent get('bg');
    }
  }

  &--left &__tooltip {
    padding-right: get('arrow-size') + get('space-between-arrow-and-trigger');

    &:after,
    &:before {
      border-color: transparent get('bdr-color');
      border-width: get('arrow-size') 0px get('arrow-size') get('arrow-size');
      right: get('space-between-arrow-and-trigger');
      top: 50%;
      transform: translateY(-50%);
    }

    &:after {
      right: get('space-between-arrow-and-trigger') + get('bdr-width');
      border-color: transparent get('bg');
    }
  }

  &__tooltip-content {
    position: relative;
    z-index: 0;
    overflow: auto;
    min-width: get('min-width');
    max-height: get('max-height');
    max-width: get('max-width');
    font-size: get('fs');
    box-shadow: get('shadow');
    border: solid get('bdr-width') get('bdr-color');
    background-color: get('bg');
    border-radius: get('bdr-radius');
  }

  // transparent zone that superpose to the trigger
  // used to keep hover state when displaying
  &__ghost-zone {
    pointer-events: initial;
    position: absolute;
  }
}
