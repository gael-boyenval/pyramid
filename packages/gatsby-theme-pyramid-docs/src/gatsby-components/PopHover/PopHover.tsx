import React, { useRef, useState, useEffect } from 'react';
import { createPortal } from 'react-dom';
import cx from 'classnames';

import { computePosition, defineBestSide } from './ComputePosition';
import './PopHover.scss';

const Portal = ({ children }) => {
  const mount = document.getElementById('pop-hover');
  const el = document.createElement('div');

  useEffect(() => {
    mount.appendChild(el);
    return () => mount.removeChild(el);
  }, [el, mount]);

  return createPortal(children, el);
};

const PopHover: React.FC<PopHoverProps> = ({
  children,
  Infos,
  popSide = 'auto',
}): JSX.Element => {
  const content = useRef(null);
  const [side, setSide] = useState('top');
  const [elementsPosition, setElementsPosition] = useState({});
  const [isRootHover, setIsRootHover] = useState(false);
  const [isPopHover, setIsPopHover] = useState(false);
  const [isVisible, setIsVisible] = useState(false);
  const [position, setPosition] = useState(undefined);

  useEffect(() => {
    if (!isVisible) return;

    const handleKeyDown = (event): void => {
      event = event || window.event;
      // escape pressed => make sure to hide the tooltip
      if (event.keyCode == 27) {
        setIsVisible(false);
      }
    };

    window.addEventListener('keydown', handleKeyDown);

    return () => {
      window.removeEventListener('keydown', handleKeyDown, true);
    };
  }, [isVisible]);

  // compute styles position depending on available
  // either the source component or the tooltip
  useEffect(() => {
    if (!position) return;
    popSide === 'auto'
      ? setSide(defineBestSide(position, window.innerHeight, window.innerWidth))
      : setSide(popSide);

    setElementsPosition(
      computePosition(position, window.innerHeight, window.innerWidth),
    );

    setIsVisible(isPopHover || isRootHover);
  }, [position]);

  // add bounding positions of trigger element
  // to state when component mount
  useEffect(() => {
    setPosition(content.current.getBoundingClientRect());
  }, []);

  // trigger position computations when hovering
  // either the source component or the tooltip
  useEffect(() => {
    setPosition(content.current.getBoundingClientRect());
  }, [isRootHover, isPopHover]);

  return Infos ? (
    <>
      <div
        onMouseOver={() => setIsRootHover(true)}
        onFocus={() => setIsRootHover(true)}
        onBlur={() => setIsRootHover(false)}
        ref={content}
        className="c-pop-hover-wrapper"
        tabIndex={0}
        aria-describedby="tooltip"
      >
        {children}
      </div>
      {isVisible && position && (
        <Portal>
          <div
            id="tooltip"
            role="tooltip"
            onMouseOver={() => setIsPopHover(true)}
            onMouseLeave={() => {
              setIsPopHover(false);
              setIsRootHover(false);
            }}
            className={cx(`c-pop-hover c-pop-hover--${side}`, {
              'is-visible': isVisible,
            })}
            style={elementsPosition[side].global}
          >
            <div
              className="c-pop-hover__ghost-zone"
              style={elementsPosition[side].ghost}
            ></div>

            <div className="c-pop-hover__infos-wrapper">
              <div className="c-pop-hover__infos">
                <Infos />
              </div>
            </div>
          </div>
        </Portal>
      )}
    </>
  ) : (
    <>{children}</>
  );
};

interface PopHoverProps {
  children: React.ReactNode;
  Infos?: React.ReactNode;
  popSide?: 'auto' | 'top' | 'bottom' | 'left' | 'right';
}

PopHover.displayName = 'PopHover';

export default PopHover;
