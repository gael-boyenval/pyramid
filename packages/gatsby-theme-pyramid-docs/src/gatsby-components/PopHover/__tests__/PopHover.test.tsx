import React from 'react';
import { mount } from 'enzyme';
import PopHover from '../PopHover';
import * as ComputePosition from '../ComputePosition';
import toJson from 'enzyme-to-json';

describe('<PopHover />', () => {
  let component;

  // add a div with #modal-root id to the global body
  const popHoverRoot = global.document.createElement('div');
  popHoverRoot.setAttribute('id', 'pop-hover');
  const body = global.document.querySelector('body');
  body.appendChild(popHoverRoot);

  afterEach(() => {
    component.unmount();
  });

  const mountPopHover = (side = 'auto') =>
    mount(
      <PopHover
        popSide={side}
        Content={() => <span className="tooltip">TOOLTIP CONTENT</span>}
      >
        <div className="trigger">TRIGGER</div>
      </PopHover>,
    );

  it('Should render the trigger only at first', () => {
    component = mountPopHover();
    expect(component.find('.trigger').exists()).toBeTruthy();
    expect(component.find('.tooltip').exists()).toBeFalsy();
    expect(toJson(component)).toMatchSnapshot();
  });

  it('Should render the tooltip when mouseover the trigger', () => {
    component = mountPopHover();
    const wrapper = component.find(PopHover);

    // hovering the trigger => show
    wrapper.simulate('mouseover');
    expect(component.find('.tooltip').exists()).toBeTruthy();
    expect(toJson(component)).toMatchSnapshot();
  });

  it('Should hide the tooltip when mouseleave the tooltip', () => {
    component = mountPopHover();
    const wrapper = component.find(PopHover);

    // hovering the trigger => show
    wrapper.simulate('mouseover');
    // mouse leave the tooltip => hide
    component.find('#tooltip').simulate('mouseleave');
    expect(component.find('#tooltip').exists()).toBeFalsy();
    expect(toJson(component)).toMatchSnapshot();
  });

  it('Should works with keyboard nav (using focus)', () => {
    component = mountPopHover();
    const wrapper = component.find(PopHover);

    // when in focus => show
    wrapper.simulate('focus');
    expect(component.find('.tooltip').exists()).toBeTruthy();

    // when blur => hide
    wrapper.simulate('blur');
    expect(component.find('.tooltip').exists()).toBeFalsy();
  });

  it('Should position on the side where there is the most available space', () => {
    const defineBestSideSpy = jest.spyOn(ComputePosition, 'defineBestSide');

    component = mountPopHover();

    // there is more available space at Right here
    global.innerWidth = 1000;
    global.innerHeight = 600;
    const wrapper = component.find(PopHover);

    const getBoundingClientRectSpy = jest.fn(() => ({
      right: 200,
      top: 100,
      bottom: 200,
      left: 100,
      width: 100,
      height: 100,
      y: 0,
      x: 0,
      toJSON: () => {},
    }));

    Element.prototype.getBoundingClientRect = getBoundingClientRectSpy;

    // trigger display to compute position
    wrapper.simulate('focus');
    expect(defineBestSideSpy).toBeCalled();
    expect(component.find('.c-pop-hover--right').exists()).toBeTruthy();
  });

  it('Should allow manually setting tooltip side', () => {
    const defineBestSideSpy = jest.spyOn(ComputePosition, 'defineBestSide');
    // add popSide="left" prop
    component = mountPopHover('left');
    const wrapper = component.find(PopHover);

    wrapper.simulate('focus');
    expect(toJson(component)).toMatchSnapshot();
    expect(defineBestSideSpy).not.toHaveBeenCalled();
    expect(component.find('.c-pop-hover--left').exists()).toBeTruthy();
    expect(component.find('.c-pop-hover--right').exists()).toBeFalsy();
  });
});
