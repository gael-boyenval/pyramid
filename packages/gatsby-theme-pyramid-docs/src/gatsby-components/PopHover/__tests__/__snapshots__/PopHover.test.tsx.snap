// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PopHover /> Should allow manually setting side 1`] = `
<PopHover
  Content={[Function]}
  popSide="left"
>
  <div
    aria-describedby="tooltip"
    className="c-pop-hover-wrapper"
    onBlur={[Function]}
    onFocus={[Function]}
    onMouseOver={[Function]}
    tabIndex={0}
  >
    <div
      className="trigger"
    >
      TRIGGER
    </div>
  </div>
  <Portal>
    <Portal
      containerInfo={
        <div>
          <div
            class="c-pop-hover c-pop-hover--left is-visible"
            id="tooltip"
            role="tooltip"
            style="top: 150px; transform: translateY(-50%); right: 800px; padding-right: 100px;"
          >
            <div
              class="c-pop-hover__ghost-zone"
              style="top: 50%; right: 0px; transform: translateY(-50%); width: 130px; height: 120px;"
            />
            <div
              class="c-pop-hover__tooltip"
            >
              <div
                class="c-pop-hover__tooltip-content"
              >
                <span
                  class="tooltip"
                >
                  TOOLTIP CONTENT
                </span>
              </div>
            </div>
          </div>
        </div>
      }
    >
      <div
        className="c-pop-hover c-pop-hover--left is-visible"
        id="tooltip"
        onMouseLeave={[Function]}
        onMouseOver={[Function]}
        role="tooltip"
        style={
          Object {
            "paddingRight": 100,
            "right": 800,
            "top": 150,
            "transform": "translateY(-50%)",
          }
        }
      >
        <div
          className="c-pop-hover__ghost-zone"
          style={
            Object {
              "height": 120,
              "right": 0,
              "top": "50%",
              "transform": "translateY(-50%)",
              "width": 130,
            }
          }
        />
        <div
          className="c-pop-hover__tooltip"
        >
          <div
            className="c-pop-hover__tooltip-content"
          >
            <Content>
              <span
                className="tooltip"
              >
                TOOLTIP CONTENT
              </span>
            </Content>
          </div>
        </div>
      </div>
    </Portal>
  </Portal>
</PopHover>
`;

exports[`<PopHover /> Should render hide the tooltip when mouseleave the tooltip 1`] = `
<PopHover
  Content={[Function]}
  popSide="auto"
>
  <div
    aria-describedby="tooltip"
    className="c-pop-hover-wrapper"
    onBlur={[Function]}
    onFocus={[Function]}
    onMouseOver={[Function]}
    tabIndex={0}
  >
    <div
      className="trigger"
    >
      TRIGGER
    </div>
  </div>
</PopHover>
`;

exports[`<PopHover /> Should render the tooltip when mouseover the trigger 1`] = `
<PopHover
  Content={[Function]}
  popSide="auto"
>
  <div
    aria-describedby="tooltip"
    className="c-pop-hover-wrapper"
    onBlur={[Function]}
    onFocus={[Function]}
    onMouseOver={[Function]}
    tabIndex={0}
  >
    <div
      className="trigger"
    >
      TRIGGER
    </div>
  </div>
  <Portal>
    <Portal
      containerInfo={
        <div>
          <div
            class="c-pop-hover c-pop-hover--right is-visible"
            id="tooltip"
            role="tooltip"
            style="top: 0px; transform: translateY(-50%); left: 0px; padding-left: 0px;"
          >
            <div
              class="c-pop-hover__ghost-zone"
              style="top: 50%; left: 0px; transform: translateY(-50%); width: 30px; height: 20px;"
            />
            <div
              class="c-pop-hover__tooltip"
            >
              <div
                class="c-pop-hover__tooltip-content"
              >
                <span
                  class="tooltip"
                >
                  TOOLTIP CONTENT
                </span>
              </div>
            </div>
          </div>
        </div>
      }
    >
      <div
        className="c-pop-hover c-pop-hover--right is-visible"
        id="tooltip"
        onMouseLeave={[Function]}
        onMouseOver={[Function]}
        role="tooltip"
        style={
          Object {
            "left": 0,
            "paddingLeft": 0,
            "top": 0,
            "transform": "translateY(-50%)",
          }
        }
      >
        <div
          className="c-pop-hover__ghost-zone"
          style={
            Object {
              "height": 20,
              "left": 0,
              "top": "50%",
              "transform": "translateY(-50%)",
              "width": 30,
            }
          }
        />
        <div
          className="c-pop-hover__tooltip"
        >
          <div
            className="c-pop-hover__tooltip-content"
          >
            <Content>
              <span
                className="tooltip"
              >
                TOOLTIP CONTENT
              </span>
            </Content>
          </div>
        </div>
      </div>
    </Portal>
  </Portal>
</PopHover>
`;

exports[`<PopHover /> Should render the trigger only at first 1`] = `
<PopHover
  Content={[Function]}
  popSide="auto"
>
  <div
    aria-describedby="tooltip"
    className="c-pop-hover-wrapper"
    onBlur={[Function]}
    onFocus={[Function]}
    onMouseOver={[Function]}
    tabIndex={0}
  >
    <div
      className="trigger"
    >
      TRIGGER
    </div>
  </div>
</PopHover>
`;
